pkg/cmd/roachtest/operations/resize.go

```diff
@@ -0,0 +1,142 @@
+// replaces the listen IP and node specific arguments with a new IP and node ID.
+// This is required because the operation has no knowledge of the original start
+// options and the best we can currently do is to copy it from the first node
+// and modify the relevant parts for the new node.
+func transformStartScript(startScript, newIP string, newNodeID int) string {
+	newLine := fmt.Sprintf("--listen-addr=%s:", newIP)
+	output := advertiseAddrRegex.ReplaceAllString(startScript, newLine)
+	output = roachprodRegex.ReplaceAllString(output, fmt.Sprintf("ROACHPROD=%d", newNodeID))
+	return nodeIDRegex.ReplaceAllString(output, fmt.Sprintf("node%d", newNodeID))
+}
+
+type cleanupResize struct {
+	growCount int
+}
+
+// Cleanup shrinks the cluster back to its original size.
+func (cl *cleanupResize) Cleanup(ctx context.Context, o operation.Operation, c cluster.Cluster) {
+	err := c.Shrink(ctx, o.L(), cl.growCount)
+	if err != nil {
+		o.Status(fmt.Sprintf("error shrinking cluster %s", err))
+	} else {
+		o.Status("shrunk cluster back to original size")
+	}
+}
+
+// resizeCluster grows the cluster by the specified count, waits for a specified
+// duration, then drains and decommissions the new nodes.
+func resizeCluster(
+	ctx context.Context,
+	o operation.Operation,
+	c cluster.Cluster,
+	growCount int,
+	pauseDuration time.Duration,
+) *cleanupResize {
+	// Create a temporary directory to store the required files for this operation.
+	tmpDir, err := os.MkdirTemp("", "")
+	if err != nil {
+		o.Fatal(err)
+	}
+	defer func() { _ = os.RemoveAll(tmpDir) }()
+
+	// Grab required files from the first node.
+	startScript, err := getStartScript(ctx, c)
+	if err != nil {
+		o.Fatal(err)
+	}
+	err = c.Get(ctx, o.L(), "cockroach", path.Join(tmpDir, "cockroach"), c.Node(1))
+	if err != nil {
+		o.Fatal(errors.Wrap(err, "failed to get cockroach, please ensure cockroach is running on this cluster."))
+	}
+	err = c.Get(ctx, o.L(), "lib", path.Join(tmpDir, "lib"), c.Node(1))
+	if err != nil {
+		o.Fatal(errors.Wrap(err, "failed to get cockroach libs."))
+	}
+
+	// Grow the cluster, but keep track of the original cluster size.
+	origClusterSize := c.Spec().NodeCount
+	err = c.Grow(ctx, o.L(), growCount)
+	if err != nil {
+		o.Fatal(err)
+	}
+	newNodes := c.Range(origClusterSize+1, origClusterSize+growCount)
+
+	// Copy the required files to the new nodes.
+	c.Put(ctx, path.Join(tmpDir, "cockroach"), "cockroach", newNodes)
+	c.Put(ctx, path.Join(tmpDir, "lib"), "lib", newNodes)
+	newIPs, err := c.InternalIP(ctx, o.L(), newNodes)
+	if err != nil {
+		o.Fatal(err)
+	}
+	for i := 0; i < growCount; i++ {
+		newStartScript := transformStartScript(startScript, newIPs[i], origClusterSize+i+1)
+		err = c.PutString(ctx, newStartScript, "cockroach.sh", 0700, c.Node(origClusterSize+i+1))
+		if err != nil {
+			o.Fatal(err)
+		}
+	}
+
+	// Start the new nodes.
+	c.Run(ctx, option.WithNodes(newNodes), "./cockroach.sh")
```

Comment:
For my own understanding, why is it necessary to go through the trouble of editing the startup script (and coupling ourselves to the current implementation) -- what breaks if we use `c.Start` on the new nodes?

Reply:
We're running into our "roachprod keeps no state" limitation here. In the case of DRT some clusters have additional args passed with the `start` command. We have no way of knowing what those were. One example is the store count, which gets passed. If we were to just invoke start on the new nodes it would not match the rest of the cluster.

**An alternate here may be that we need to supply some yaml / config with an operation runner** so it has some insight about what it is actually running operations against. I see a few new operation on the horizon that look for "tpcc" or make some assumption and then bail out if it is not true. @itsbilal Any thoughts on this?

Reply:
One update I could make here is to use `Start with Restart`, which essentially just runs the script as well, but we'd still need to either copy & mutate a script from an existing node in the cluster, or store the initial start flags somewhere (could either be passed as flags to run operations, or a config file) - or we could delegate to `drt-run` to pass the flags from a config?

Reply:
TLDR; I think I'm rather going to expose `StartOpts` as a flag for running operations. In the future the runner of the operation will have to know the details of how the cluster was started.